domains/client/cross-domain-message-gossip/src/message_listener.rs

```diff
@@ -38,46 +119,378 @@ pub async fn start_cross_chain_message_listener<Client, TxPool, TxnListener>(
+    Executor: CodeExecutor + RuntimeVersionOf,
+{
+    let runtime_api = consensus_client.runtime_api();
+    let consensus_best_hash = consensus_client.info().best_hash;
+    let consensus_block_header = consensus_client
+        .header(consensus_best_hash)?
+        .ok_or(Error::MissingBlockHeader)?;
+
+    let domain_runtime_type = runtime_api
+        .domain_instance_data(*consensus_block_header.parent_hash(), src_domain_id)?
+        .ok_or(Error::MissingDomainRuntimeCode)?
+        .0
+        .runtime_type;
+
+    let is_valid_domain_block_number =
+        |block_number: BlockNumber| -> Result<(Block::Hash, Block::Hash), Error> {
+            let receipt_hash = runtime_api
+                .receipt_hash(consensus_best_hash, src_domain_id, block_number.into())?
+                .ok_or(Error::MissingDomainReceiptHash)?;
+
+            // check if the receipt is challenged by fraud proof
+            if runtime_api.is_bad_er_pending_to_prune(
+                consensus_best_hash,
+                src_domain_id,
+                receipt_hash,
+            )? {
+                return Err(Error::BadDomainReceiptHash);
+            }
+
+            let receipt = runtime_api
+                .execution_receipt(consensus_best_hash, receipt_hash)?
+                .ok_or(Error::MissingDomainReceipt)?;
+
+            Ok((receipt.domain_block_hash, receipt.final_state_root))
+        };
+
+    // check if the domain block number is valid
+    let (domain_block_hash, domain_state_root) = is_valid_domain_block_number(domain_block_number)?;
+
+    // if there is an existing channel detail,
+    // return if the channel update is from canonical domain and block number is latest
+    // else store the update.
+    let maybe_existing_channel_detail = crate::aux_schema::get_channel_details(
+        &**consensus_client,
+        ChainId::Domain(src_domain_id),
+        dst_chain_id,
+        channel_id,
+    )?;
+    if let Some(existing_channel_update) = maybe_existing_channel_detail {
+        // if the existing update domain block number
+        // is valid, and
+        // more the new block number, then don't update
+        if let Ok((existing_block_hash, _)) =
+            is_valid_domain_block_number(existing_channel_update.block_number)
+        {
+            if existing_block_hash.as_ref() == existing_channel_update.block_hash.as_ref()
+                && existing_channel_update.block_number >= domain_block_number
+            {
+                return Ok(());
+            }
+        }
+    }
+
+    let domain_runtime = runtime_api
+        .domain_runtime_code(*consensus_block_header.parent_hash(), src_domain_id)?
+        .ok_or(Error::MissingDomainRuntimeCode)?;
+    let domain_stateless_runtime =
+        StatelessRuntime::<Block, _>::new(executor.clone(), domain_runtime.into());
+    let storage_key =
+        StorageKey(domain_stateless_runtime.channel_storage_key(dst_chain_id, channel_id)?);
```

Comment:
How about maintaining a cache for the storage key `Map<(runtime_hash, chain_id, channel_id), StorageKey>` here? so we don't need to create a runtime instance for every message.

Reply:
Make sense!